janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java

```diff
@@ -737,87 +754,29 @@ public void mutate(Map<String, Map<String, IndexMutation>> mutations, KeyInforma
-                        .append(jsValue)
-                        .append("); ctx._source[\"")
-                        .append(deletion.field)
-                        .append("\"].remove(")
-                        .append(index)
-                        .append(");");
-                    if (hasDualStringMapping(information.get(storeName, deletion.field))) {
-                        index = INDEX_NAME + i++;
-                        script.append("def ")
-                            .append(index).append(" = ctx._source[\"")
-                            .append(getDualMappingName(deletion.field))
-                            .append("\"].indexOf(")
-                            .append(jsValue)
-                            .append("); ctx._source[\"")
-                            .append(getDualMappingName(deletion.field))
-                            .append("\"].remove(")
-                            .append(index)
-                            .append(");");
-                    }
-                    break;
-            }
-        }
-        return script.toString();
-    }
-
-    private String getAdditionScript(KeyInformation.IndexRetriever information, String storeName,
-                                     IndexMutation mutation) throws PermanentBackendException {
-        final StringBuilder script = new StringBuilder();
-
-        for (final IndexEntry e : mutation.getAdditions()) {
-            final KeyInformation keyInformation = information.get(storeName).get(e.field);
-            final Cardinality cardinality = keyInformation.getCardinality();
-
-            if (cardinality != Cardinality.SET && cardinality != Cardinality.LIST) {
+    private List<Map<String, Object>> getParameters(KeyInformation.StoreRetriever storeRetriever,
+                                                    List<IndexEntry> entries,
+                                                    boolean deletion,
+                                                    Cardinality... cardinalitiesToSkip) {
+        Set<Cardinality> cardinalityToSkipSet = Sets.newHashSet(cardinalitiesToSkip);
+        List<Map<String, Object>> result = new ArrayList<>();
+        for (IndexEntry entry : entries) {
+            KeyInformation info = storeRetriever.get(entry.field);
+            if (cardinalityToSkipSet.contains(info.getCardinality())) {
                 continue;
             }
-
-            String value = convertToJsType(e.value, compat.scriptLang(), Mapping.getMapping(keyInformation));
-
-            appendAdditionScript(script, value, e.field, cardinality == Cardinality.SET);
-
-            if (hasDualStringMapping(keyInformation)) {
-                appendAdditionScript(script, value, getDualMappingName(e.field), cardinality == Cardinality.SET);
+            Object jsValue = deletion && info.getCardinality() == Cardinality.SINGLE ?
+                    entry.value : convertToEsType(entry.value, Mapping.getMapping(info));
```

Comment:
This is not necessary, but if `cardinality` == SINGLE for deletion, we don't need `entry.value` at all. So we could just make:
```
Object jsValue = deletion && info.getCardinality() == Cardinality.SINGLE ?
                    "" : convertToEsType(entry.value, Mapping.getMapping(info));
```

Reply:
done